istioctl/cmd/sidecar-bootstrap.go

```diff
@@ -0,0 +1,643 @@
+		_, err = fmt.Fprintln(w, "C0644", size, filename)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		if err = checkRemoteResponse(stdout); err != nil {
+			errCh <- err
+			return
+		}
+
+		_, err = io.Copy(w, r)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		_, err = fmt.Fprint(w, "\x00")
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		if err = checkRemoteResponse(stdout); err != nil {
+			errCh <- err
+			return
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		err := session.Run(fmt.Sprintf("%s -qt %s", scpPath, location))
+		if err != nil {
+			errCh <- err
+			return
+		}
+	}()
+
+	if waitTimeout(&wg, scpTimeout) {
+		return fmt.Errorf("timeout uploading file")
+	}
+
+	close(errCh)
+	for err := range errCh {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func remoteRunCommand(command string, client *ssh.Client) error {
+	session, err := client.NewSession()
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+	return session.Run(command)
+}
+
+func dumpCertificates(directory string, addressCertMapping map[string]workloadEntryAddressKeys) error {
+	err := os.MkdirAll(directory, os.ModePerm)
+	if err != nil && !os.IsExist(err) {
+		return err
+	}
+	for address, certs := range addressCertMapping {
+		err = ioutil.WriteFile(path.Join(directory, "cert-"+address+".pem"), certs.Cert, 0644)
+		if err != nil {
+			return err
+		}
+		err = ioutil.WriteFile(path.Join(directory, "key-"+address+".pem"), certs.Key, 0644)
+		if err != nil {
+			return err
+		}
+		err = ioutil.WriteFile(path.Join(directory, "ca-cert-"+address+".pem"), certs.CaCert, 0644)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func addressToPodNameAddition(address string) string {
+	return fmt.Sprintf("%x", sha256.Sum256([]byte(address)))[0:7]
+}
+
+func deriveDiscoveryAddressPort(service string, namespace string, kubeClient kubernetes.Interface) (string, error) {
+	foundService, err := kubeClient.CoreV1().Services(namespace).Get(context.TODO(), service, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+	for _, port := range foundService.Spec.Ports {
+		if strings.HasPrefix(port.Name, "tcp-istiod") {
+			return string(port.NodePort), nil
+		}
+	}
+	return "", fmt.Errorf("failed to find tcp-istiod port")
+}
+
+func copyCertificates(kubeClient kubernetes.Interface,
+	addressCertMapping map[string]workloadEntryAddressKeys) error {
+	var callback ssh.HostKeyCallback
+	if sshIgnoreHostKeys {
+		callback = ssh.InsecureIgnoreHostKey()
+	} else {
+		user, err := user.Current()
+		if err != nil {
+			return err
+		}
+		callback, err = knownhosts.New(path.Join(user.HomeDir, ".ssh", "known_hosts"))
+		if err != nil {
+			return err
+		}
+	}
+
+	for address, certs := range addressCertMapping {
+		sshConfig := &ssh.ClientConfig{
+			User:            sshUser,
+			Auth:            []ssh.AuthMethod{sshAuthMethod},
+			HostKeyCallback: callback,
+		}
+		client, err := ssh.Dial("tcp", address+":22", sshConfig)
```

Comment:
I can't think if an easy way to support SSH to non-default port, but if you can maybe we should allow it?